ext/dom/php_dom.c

```diff
@@ -470,17 +511,56 @@ PHP_FUNCTION(dom_import_simplexml)
 	nodeobj = (php_libxml_node_object *) ((char *) Z_OBJ_P(node) - Z_OBJ_HT_P(node)->offset);
 	nodep = php_libxml_import_node(node);
 
-	if (nodep && nodeobj && (nodep->type == XML_ELEMENT_NODE || nodep->type == XML_ATTRIBUTE_NODE)) {
+	if (nodep && nodeobj && nodeobj->document && (nodep->type == XML_ELEMENT_NODE || nodep->type == XML_ATTRIBUTE_NODE)) {
+		php_libxml_class_type old_class_type = nodeobj->document->class_type;
+		if (old_class_type != PHP_LIBXML_CLASS_UNSET && old_class_type != new_class) {
+			if (new_class == PHP_LIBXML_CLASS_MODERN) {
+				zend_argument_type_error(1, "must not be already imported as a DOMNode");
+			} else {
+				zend_argument_type_error(1, "must not be already imported as a DOM\\Node");
+			}			
+			RETURN_THROWS();
+		}
+
+		/* Lock the node class type to prevent creating multiple representations of the same node. */
+		nodeobj->document->class_type = new_class;
+
+		if (old_class_type != PHP_LIBXML_CLASS_MODERN && new_class == PHP_LIBXML_CLASS_MODERN && nodep->doc != NULL) {
+			dom_document_convert_to_modern(nodeobj->document, nodep->doc);
+		}
+
 		DOM_RET_OBJ((xmlNodePtr) nodep, &ret, (dom_object *)nodeobj);
 	} else {
+		/* TODO: should this be a type error instead? */
 		zend_argument_value_error(1, "is not a valid node type");
 		RETURN_THROWS();
```

Comment:
Probably yes, if the input type should be a union type of sort?

Reply:
No it's one of those `object` cases with special handling; I changed some in the past too to use a type error.

Reply:
Right, I have a vague recollection of this